### ISSUE TITLE
proto: add function to shallow copy a message

### ISSUE THREAD

**Issue by dsnet** [golang/protobuf#1155]
Generated messages have mutexes and other variables that make it dangerous to shallow copy.

Add a function to the `proto` package that does the equivalent of a shallow copy in a safe way.

The function would semantically equivalent to:
```go
func ShallowCopy(dst, src proto.Message) {
     dm := dst.ProtoReflect()
     sm := src.Protoreflect()
     if dm.Type() != sm.Type() {
         panic("mismatching type")
     }
     proto.Reset(dst)
     sm.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
         dm.Set(fd, v)
         return true
     })
}
```
but be optimized for performance.

**Comment by eccles** [comment]
I have just had to abandon an upgrade because we copy messages. Because they now contain mutexes the copylocks check of the  go vet command now fails in numerous places.
So I need this for correctness - it is a serious error to copy mutexes which explains why 'go vet' has check explicitly for when this happens.

**Comment by dsnet** [comment]
> I have just had to abandon an upgrade because we copy messages. Because they now contain mutexes the copylocks check of the go vet command now fails in numerous places.

The data structures that were not safe for shallow copying have always been present since v1.1 (and possibly earlier). The only difference is that we made it possible for vet to detect such cases more readily. In other words, just because vet didn't flag it for v1.3 and earlier doesn't mean it was okay on those versions.

**Comment by mostrows2** [comment]
> > I have just had to abandon an upgrade because we copy messages. Because they now contain mutexes the copylocks check of the go vet command now fails in numerous places.
> 
> The data structures that were not safe for shallow copying have always been present since v1.1 (and possibly earlier). The only difference is that we made it possible for vet to detect such cases more readily. In other words, just because vet didn't flag it for v1.3 and earlier doesn't mean it was okay on those versions.

It isn't obvious to users of these libraries that it wasn't safe.  The point of generated code is that a user shouldn't need to be reading it to figure out implementation subtleties.

It is good that you've made everyone aware of the issue but...  at the same time there's no way to remediate the problem.
I too am stuck now in trying to upgrade to these libraries, but am hosed because I have thousands of these errors coming up now (mainly in logging statements).

**Comment by janpfeifer** [comment]
This seems indeed an important issue. Copying (both shallow and deep) and comparison is one of the most basic operations one would want to perform -- not without reason Go supports assignment and comparison for structs.

For end-users to deal with it can be inneficient+incovenient (writing a generic function with reflection), or very inconvient (having to write a  copy/equal functions by hand for each message, as needed).

Notice that [C++ API](https://developers.google.com/protocol-buffers/docs/cpptutorial#standard-message-methods), [Java API](https://developers.google.com/protocol-buffers/docs/pythontutorial#standard-message-methods) and [Python API] all provide a `CopyFrom()` method. For equality I've seen methods for that in testing packages, but I can't find them right now.

**Comment by neild** [comment]
For future reference: You probably mean [`proto.Clone`](https://pkg.go.dev/google.golang.org/protobuf/proto#Clone), rather than `Equal`.

**Comment by eccles** [comment]
Question: proto.Clone() creates a new protobuff message from the old one. What If I do not want that? Instead I want to create a new struct that only contains the data fields without any protobuf fields (XXX and suchlike). In that case I want an explicit CopyFrom(), MergeFrom() methods ? proto.Clone() is not suitable in this case...

It might be possible to use proto.Marshal()/Unmarsha() which gets you most of the way there albeit with the protobuf fields still present but nil.?

**Comment by puellanivis** [comment]
> What If I do not want that? Instead I want to create a new struct that only contains the data fields without any protobuf fields (XXX and suchlike). In that case I want an explicit CopyFrom(), MergeFrom() methods ?

😟 I don’t think something like that is going to be easily accomplished without using protobuf reflection. How could these hypothetical `CopyFrom` and `MergeFrom` methods work against an arbitrary type like that without using reflection?

**Comment by eccles** [comment]
@puellanivis yes - I think this can only be accomplished if the generated code  creates these methods as part of the struct definition...

**Comment by eaglerayp** [comment]
+1

**Comment by joeycumines** [comment]
Mmm I found this topic because I was hoping for a better solution to shallow copy protobuf message(s). I'd guess you're right @puellanivis @eccles, in that the only way to accomplish "copy without reflection" would be to implement it manually, or generate code for it.

My use case typically involves patching part of an existing message, without actually modifying the existing value. It's been my experience that treating messages as read-only often the only sane choice, if you're passing protobuf messages around (within a single process).

This is my solution, which uses the `reflect` package. I am not sure why I didn't use protobuf reflection. I may have I run into issues, or perhaps I just didn't bother investigating it 😂.

```go
// shallow copy methods could also be implemented manually, though that's a bit of a pain

var copySomeMessage = ProtoCopier((*SomeMessage)(nil))

func (x *SomeMessage) Copy() *SomeMessage { return copySomeMessage(x).(*SomeMessage) }
```

```go
// ProtoCopier returns a function that will shallow copy values of a given protobuf value's type, utilising any `Get`
// prefixed method, accepting no input parameters, and returning a single value of the same type, available for each
// given field, intended to be be used with generated code for protobuf messages
// NOTE a panic will occur if the v's type is not t
func ProtoCopier(v interface{}) func(v interface{}) interface{} {
	var (
		t            = reflect.TypeOf(v)
		fieldMethods = make([][2]int, 0)
	)
	for i := 0; i < t.Elem().NumField(); i++ {
		field := t.Elem().Field(i)
		if field.PkgPath != `` {
			continue
		}
		method, ok := t.MethodByName(`Get` + field.Name)
		if !ok ||
			method.Type.NumIn() != 1 ||
			method.Type.NumOut() != 1 ||
			method.Type.Out(0) != field.Type {
			continue
		}
		fieldMethods = append(fieldMethods, [2]int{i, method.Index})
	}
	return func(v interface{}) interface{} {
		src := reflect.ValueOf(v)
		protoCopierCheckType(t, src.Type())
		dst := reflect.New(t.Elem()).Elem()
		for _, fieldMethod := range fieldMethods {
			dst.Field(fieldMethod[0]).Set(src.Method(fieldMethod[1]).Call(nil)[0])
		}
		return dst.Addr().Interface()
	}
}

func protoCopierCheckType(dst, src reflect.Type) {
	if dst != src {
		panic(fmt.Errorf(`ProtoCopier dst %s != src %s`, dst, src))
	}
}
```

**Comment by joeycumines** [comment]
@dsnet I'm not sure if it's what you had in mind, and I'd still love to see something added to the official plugin, but I've implemented a protoc plugin to address this use case.

https://github.com/joeycumines/protoc-gen-go-msgclone
